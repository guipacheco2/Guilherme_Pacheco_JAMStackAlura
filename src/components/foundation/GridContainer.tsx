import React from 'react'
import styled, { css, CSSProperties } from 'styled-components'
import {
  breakpointsMedia,
  propsToStyle,
  ResponsiveBreakpoints,
} from '../../theme'
import { BreakpointKeys, breakpointNames } from '../../theme/breakpoints'

export interface StyledGridContainerProps {
  alignItems?: ResponsiveBreakpoints<CSSProperties['alignItems']>
  display?: ResponsiveBreakpoints<CSSProperties['display']>
  flex?: ResponsiveBreakpoints<CSSProperties['flex']>
  marginTop?: ResponsiveBreakpoints<CSSProperties['marginTop']>
  maxWidth?: BreakpointKeys
}

const StyledGridContainer = styled.div<StyledGridContainerProps>(
  ({ alignItems, display, flex, marginTop, maxWidth = 'xl', theme }) => {
    const breakpointIndex = breakpointNames.indexOf(maxWidth)

    return css`
      width: 100%;
      margin-right: auto;
      margin-left: auto;
      ${breakpointsMedia({
        lg:
          breakpointIndex >= 3
            ? css`
                max-width: ${theme.breakpoints.lg}px;
                padding-right: 24px;
                padding-left: 24px;
              `
            : css``,
        md:
          breakpointIndex >= 2
            ? css`
                max-width: ${theme.breakpoints.md}px;
                padding-right: 24px;
                padding-left: 24px;
              `
            : css``,
        sm:
          breakpointIndex >= 1
            ? css`
                max-width: ${theme.breakpoints.sm}px;
                padding-right: 24px;
                padding-left: 24px;
              `
            : css``,
        xl:
          breakpointIndex >= 4
            ? css`
                max-width: ${theme.breakpoints.xl}px;
                padding-right: 24px;
                padding-left: 24px;
              `
            : css``,
        xs:
          breakpointIndex >= 0
            ? css`
                max-width: initial;
                padding-right: 16px;
                padding-left: 16px;
              `
            : css``,
      })}
      ${propsToStyle({ alignItems, display, flex, marginTop })}
    `
  },
)

interface GridContainerProps extends StyledGridContainerProps {
  children: React.ReactNode
}

export function GridContainer({
  alignItems,
  children,
  display,
  flex,
  marginTop,
  maxWidth,
}: GridContainerProps): JSX.Element {
  return (
    <StyledGridContainer
      alignItems={alignItems}
      flex={flex}
      maxWidth={maxWidth}
      display={display}
      marginTop={marginTop}
    >
      {children}
    </StyledGridContainer>
  )
}
