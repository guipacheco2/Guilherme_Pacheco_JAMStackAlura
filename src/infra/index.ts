export * from './hooks'
export * from './test'
