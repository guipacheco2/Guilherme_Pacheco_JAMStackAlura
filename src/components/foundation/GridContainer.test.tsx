import React from 'react'
import { render } from '../../infra'
import { GridContainer } from './GridContainer'

describe('<GridContainer />', () => {
  test('renders component', () => {
    const { baseElement } = render(<GridContainer>child</GridContainer>)
    expect(baseElement).toMatchSnapshot()
  })
})
