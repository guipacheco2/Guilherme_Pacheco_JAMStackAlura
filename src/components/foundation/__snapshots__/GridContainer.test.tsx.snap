// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GridContainer /> renders component 1`] = `
.c0 {
  width: 100%;
  margin-right: auto;
  margin-left: auto;
}

@media only screen and (min-width:0px) {
  .c0 {
    max-width: initial;
    padding-right: 16px;
    padding-left: 16px;
  }
}

@media only screen and (min-width:600px) {
  .c0 {
    max-width: 600px;
    padding-right: 24px;
    padding-left: 24px;
  }
}

@media only screen and (min-width:1024px) {
  .c0 {
    max-width: 1024px;
    padding-right: 24px;
    padding-left: 24px;
  }
}

@media only screen and (min-width:1440px) {
  .c0 {
    max-width: 1440px;
    padding-right: 24px;
    padding-left: 24px;
  }
}

@media only screen and (min-width:1920px) {
  .c0 {
    max-width: 1920px;
    padding-right: 24px;
    padding-left: 24px;
  }
}

<body>
  <div>
    <div
      class="c0"
    >
      child
    </div>
  </div>
</body>
`;
